### PR TITLE
[vcpkg] Add experimental $X_VCPKG_ASSET_SOURCES for source caching

### PR DESCRIPTION
**Pull request by ras0219**
Tool changes for https://github.com/microsoft/vcpkg/pull/13639.

See https://github.com/ras0219/vcpkg/blob/dev/roschuma/azblob-src/docs/users/config-environment.md#x_vcpkg_asset_sources for behavior docs.

The `X_VCPKG_ASSET_SOURCES` is designed to be similar to `X_VCPKG_BINARY_SOURCES` with the intent that they can share backends in the future.

### CODE REVIEW

File: include/vcpkg/binarycaching.private.h [hunk 1]
```diff
@@ -3,6 +3,7 @@
 #include <vcpkg/fwd/packagespec.h>
 #include <vcpkg/fwd/vcpkgpaths.h>
 
+#include <vcpkg/base/parse.h>
```
Comment (ras0219) [inline include/vcpkg/binarycaching.private.h:6]:
```suggestion
```

File: src/vcpkg/base/files.cpp [hunk 2]
```diff
@@ -582,6 +582,25 @@ namespace vcpkg::Files
         return result;
     }
 
+    void Filesystem::create_best_link(const fs::path& to, const fs::path& from, std::error_code& ec)
+    {
+        this->create_hard_link(to, from, ec);
```
Comment (strega-nil) [inline src/vcpkg/base/files.cpp:587]: why do we want a hard link here? it seems like we should prefer a symlink, at least to me.

File: include/vcpkg/vcpkgcmdarguments.h [hunk 3]
```diff
@@ -180,6 +180,9 @@ namespace vcpkg
         constexpr static StringLiteral JSON_SWITCH = "x-json";
         Optional<bool> json = nullopt;
 
+        constexpr static StringLiteral READWRITE_MIRROR_URL_TEMPLATE_ENV = "X_VCPKG_ASSET_SOURCES";
```
Comment (ras0219) [inline include/vcpkg/vcpkgcmdarguments.h:183]:
```suggestion
        constexpr static StringLiteral ASSET_SOURCES_ENV = "X_VCPKG_ASSET_SOURCES";
```

File: src/vcpkg/base/files.cpp (hunk 2 above)
Comment (ras0219) [reply src/vcpkg/base/files.cpp:587]: This is actually old code from a previous implementation and can be simply removed -- the reason to prefer a hard link is that it does not become invalidated by move or remove

**Review by vicroms — APPROVED** [review]

File: src/vcpkg/commands.cpp [hunk 4]
```diff
@@ -98,6 +100,7 @@ namespace vcpkg::Commands
             {"x-package-info", &info},
             {"x-history", &porthistory},
             {"x-vsinstances", &vsinstances},
+            {"x-download", &xdownload},
```
Comment (dan-shaw) [inline src/vcpkg/commands.cpp:103]: Should we also add this command to `vcpkg help`?

File: src/vcpkg/base/downloads.cpp [hunk 5]
```diff
@@ -418,9 +467,11 @@ namespace vcpkg::Downloads
...
-        fs.remove(download_path, ignore_errors);
-        fs.remove(download_path_part_path, ignore_errors);
+#if defined(_WIN32)
+        download_path_part_path += fs::u8path(Strings::concat(".", _getpid(), ".part"));
+#else
+        download_path_part_path += fs::u8path(Strings::concat(".", getpid(), ".part"));
+#endif
 
         std::string errors;
         for (const std::string& url : urls)
```
Comment (dan-shaw) [inline src/vcpkg/base/downloads.cpp:477]: Maybe not relevant since this is not part of the PR. If a url is not valid, we may get an unhelpful error message. For example, in line 480 (expand below): `split_uri_view(url)` will print `Value is null` and exit if any url is missing a colon.

**Review by dan-shaw — APPROVED** [review]
